test/e2e/util.go

```diff
@@ -0,0 +1,48 @@
+package e2e
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/onsi/ginkgo/v2"
+)
+
+// HandleTestCaseFailure is responsible for attempting to collect relevant
+// testing artifacts when individual test cases fail. In the case that
+// a test passes, then this function is a no-op and will return a nil error.
+func HandleTestCaseFailure() error {
+	currentTest := ginkgo.CurrentSpecReport()
+	if !currentTest.Failed() {
+		return nil
+	}
+
+	// current test case failed. attempt to collect CI artifacts if the
+	// $ARTIFACT_DIR environment variable has been set. This variable is
+	// always present in downstream CI environments.
+	artifactDir := os.Getenv("ARTIFACT_DIR")
```

Comment:
Potential follow-up: There's overlap between how we handle the $ARTIFACT_DIR environment variable in the Makefile invocation vs. how we're handling the value of that shared variable here. It could be worth consolidating this logic into a shared CLI flag, but it's unclear whether that's the correct implementation going forward.

The main problem with this approach is the potential for skew between how ginkgo handles relative paths vs. how we're handling relative paths within this internal testing code. The expectation is that downstream CI would set the value of $ARTIFACT_DIR variable to an absolute path, so we're in the clear there, but there's no guarantee that we're correctly configuring (or enforcing) the want/need for an absolute path when running the e2e suite locally during dev workflows.

I don't think anything I mentioned is blocking, but it's still something worth noting and potentially revisiting in future phase implementations.